_overviews/FAQ/initialization-order.md

```diff
@@ -36,129 +36,125 @@ Let's observe the initialization order through the Scala REPL:
     // A: hello, dad
     // B: hello, dad
     // C: hello, dad
 
-Early definitions are a bit unwieldy, there are limitations as to what can appear and what can be referenced in an early definitions block, and they don't compose as well as lazy vals: but if a lazy val is undesirable, they present another option.  They are specified in SLS 5.1.6.
+Note that abstract `lazy val`s are supported in Scala 3, but not in Scala 2.
+In Scala 2, you can define an abstract `val` or `def` instead.
 
-Note that early definitions are deprecated in Scala 2.13; they will be replaced by trait parameters in Scala 3. So, early definitions are not recommended for use if future compatibility is a concern.
+An exception during initialization of a lazy val will cause the right-hand side to be re-evaluated on the next access: see SLS 5.2.
 
-#### Use constant value definitions ####
-    abstract class A {
-      val x1: String
-      val x2: String = "mom"
+Note that using multiple lazy vals creates a new risk: cycles among lazy vals can result in a stack overflow on first access.
 
-      println("A: " + x1 + ", " + x2)
-    }
-    class B extends A {
-      val x1: String = "hello"
-      final val x3 = "goodbye"
+### Use a nested object
 
-      println("B: " + x1 + ", " + x2)
-    }
-    class C extends B {
-      override val x2: String = "dad"
+Sometimes, uninitialized state in a subclass is accessed during construction of a superclass:
 
-      println("C: " + x1 + ", " + x2)
+    class Adder {
+      var sum = 0
+      def add(x: Int): Unit = sum += x
+      add(1)
     }
-    abstract class D {
-      val c: C
-      val x3 = c.x3   // no exceptions!
-      println("D: " + c + " but " + x3)
+    class LogAdder extends Adder {
+      private var added: Set[Int] = Set.empty
+      override def add(x: Int): Unit = { added += x; super.add(x) }
     }
-    class E extends D {
-      val c = new C
-      println(s"E: ${c.x1}, ${c.x2}, and $x3...")
+
+In this case the state can be initialized on demand by wrapping it into a local object:
+
+    class Adder {
+      var sum = 0
+      def add(x: Int): Unit = sum += x
+      add(1)
     }
-    //scala> new E
-    //D: null but goodbye
-    //A: null, null
-    //B: hello, null
-    //C: hello, dad
-    //E: hello, dad, and goodbye...
+    class LogAdder extends Adder {
+      private object state {
+        var added: Set[Int] = Set.empty
+      }
+      import state._
+      override def add(x: Int): Unit = { added += x; super.add(x) }
+    }
+
+### Early definitions: deprecated
 
-Sometimes all you need from an interface is a compile-time constant.
+Scala 2 supports early definitinos, but they are deprecated in Scala 2.13 and unsupported in Scala 3.
+See the [migration guide](https://docs.scala-lang.org/scala3/guides/migration/incompat-dropped-features.html#early-initializer) for more information.
 
-Constant values are stricter than strict and earlier than early definitions and have even more limitations,
```

Comment:
Omitting compile-time mechanisms to work around runtime issues seems important. Probably more is available in Scala 3 besides final vals.

Let's also respect that this was the original one-question FAQ. Maybe a minute of silence?